File: RxEpics/python/examples/pv_pipeline.py
"""Demonstrates a fluent read-transform-write pipeline on EPICS test PVs.

The pipeline:
  1. Read   TEST:CALC     — raw random value from the IOC
  2. Map    |x| * 2 + 1.5 — apply calibration in client
  3. Write  TEST:DOUBLE   — store the calibrated double
  4. Format as string     — pure Python string formatting
  5. Write  TEST:STRING   — store the formatted string on the IOC
  6. Read   TEST:STRING   — read back to confirm

Each step's result feeds directly into the next.
No threads. No callbacks. No intermediate variables.

Usage:
    python pv_pipeline.py

(uses test PVs: TEST:CALC, TEST:DOUBLE, TEST:STRING)
"""

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import reactivex as rx
import reactivex.operators as ops
from reactivex.scheduler.eventloop import AsyncIOScheduler
from caproto.asyncio.client import Context

from rxepics.channel import read_pv
from rxepics.channel_write import write_pv


async def main():
    loop = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)
    ctx = Context()

    done = asyncio.Event()
    result_holder: list = []

    # Step 1: read raw value
    read_pv("TEST:CALC", ctx).pipe(
        ops.do_action(on_next=lambda v: print(f"  [1] read    TEST:CALC     =  {v}")),

        # Step 2: apply calibration — pure function, no I/O
        ops.map(lambda v: abs(v) * 2.0 + 1.5),
        ops.do_action(on_next=lambda v: print(f"  [2] calibrated            =  {v}")),

        # Step 3: write calibrated double to TEST:DOUBLE
        ops.flat_map(lambda v: write_pv("TEST:DOUBLE", v, ctx)),
        ops.do_action(on_next=lambda v: print(f"  [3] wrote   TEST:DOUBLE   =  {v}")),

        # Step 4: format as string — pure function
        ops.map(lambda v: f"calibrated={v:.4f}"),
        ops.do_action(on_next=lambda v: print(f"  [4] formatted             =  {v!r}")),

        # Step 5: write string to TEST:STRING
        ops.flat_map(lambda v: write_pv("TEST:STRING", v, ctx)),
        ops.do_action(on_next=lambda v: print(f"  [5] wrote   TEST:STRING   =  {v!r}")),

        # Step 6: read back to confirm what the IOC actually has
        ops.flat_map(lambda _: read_pv("TEST:STRING", ctx)),
    ).subscribe(
        on_next=lambda v: result_holder.append(v),
        on_error=lambda e: (print(f"ERROR: {e}", file=sys.stderr), done.set()),
        on_completed=done.set,
        scheduler=scheduler,
    )

    await done.wait()

    if result_holder:
        print(f"\n  Pipeline complete.")
        print(f"  Confirmed on IOC: {result_holder[0]!r}")


if __name__ == "__main__":
    asyncio.run(main())

File: RxEpics/cpp/examples/pv_pipeline.cpp
/**
 * Fluent EpicsClient pipeline — read → map → write → read back.
 *
 * The showstopper demo for RxEpics/cpp.
 * Mirrors Python's pv_pipeline.py and rxtango pipeline.cpp.
 *
 * Usage:
 *   EPICS_PVA_ADDR_LIST=localhost ./pv_pipeline [src_pv] [dst_pv]
 *   defaults: TEST:CALC  TEST:DOUBLE
 */

#include <any>
#include <cmath>
#include <condition_variable>
#include <iostream>
#include <mutex>
#include <string>

#include <rxepics/rxepics.hpp>

int main(int argc, char* argv[]) {
    const std::string src_pv = argc > 1 ? argv[1] : "TEST:CALC";
    const std::string dst_pv = argc > 2 ? argv[2] : "TEST:DOUBLE";

    std::cout << "EpicsClient pipeline: " << src_pv << " → calibrate → " << dst_pv << "\n\n";

    std::mutex              m;
    std::condition_variable cv;
    bool                    done = false;

    rxepics::EpicsClient()
        .read(src_pv)
        .map([src_pv](std::any v) -> std::any {
            double d = std::any_cast<double>(v);
            std::cout << "  [1] read    " << src_pv << " = " << d << "\n";
            return d;
        })
        .map([](std::any v) -> std::any {
            double calibrated = std::abs(std::any_cast<double>(v)) * 2.0 + 1.5;
            std::cout << "  [2] calibrated             = " << calibrated << "\n";
            return calibrated;
        })
        .write(dst_pv)
        .map([dst_pv](std::any v) -> std::any {
            std::cout << "  [3] wrote  " << dst_pv << " = " << std::any_cast<double>(v) << "\n";
            return v;
        })
        .read(dst_pv)
        .subscribe(
            [](std::any v) {
                std::cout << "\n  Confirmed on device: " << std::any_cast<double>(v) << "\n";
            },
            [&](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
                std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one();
            },
            [&]() {
                std::cout << "  Pipeline complete.\n";
                std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one();
            }
        );

    std::unique_lock<std::mutex> lk(m);
    cv.wait_for(lk, std::chrono::seconds(15), [&]{ return done; });
    return 0;
}
